apis/core/v1alpha2/featuregate_webhook.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureGate) ValidateUpdate(old runtime.Object) error {
+	featuregatelog.Info("validate update", "name", r.Name)
+
+	oldObj, ok := old.(*FeatureGate)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected FeatureGate object, but got object of type %T", old))
+	}
+	if oldObj == nil {
+		return nil
+	}
+
+	c, err := r.getClient()
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	ctx := context.Background()
+	var allErrors field.ErrorList
+
+	allErrors = append(allErrors, r.validateFeatureExistence(ctx, c)...)
+	allErrors = append(allErrors, r.validateFeatureConflicts(ctx, c)...)
+	allErrors = append(allErrors, r.validateStabilityPolicyOverride(oldObj)...)
+	allErrors = append(allErrors, r.validateFeatureForStabilityPolicyViolation(ctx, c)...)
+
+	if len(allErrors) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("FeatureGate").GroupKind(), r.Name, allErrors)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureGate) ValidateDelete() error {
+	featuregatelog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+// validateFeatureForStabilityPolicyViolation validates the feature for any stability policy violation
+func (r *FeatureGate) validateFeatureForStabilityPolicyViolation(ctx context.Context, c client.Client) field.ErrorList {
+	var allErrors field.ErrorList
+
+	features := &FeatureList{}
+	if err := c.List(ctx, features); err != nil {
+		allErrors = append(allErrors, field.InternalError(field.NewPath("spec").Child("features"), err))
+		return allErrors
+	}
+
+	invalidFeatures := sets.String{}
+	for _, featureRef := range r.Spec.Features {
+		stabilityLevel, found := getFeatureStabilityLevel(features, featureRef.Name)
+		if !found {
+			// Feature doesn't exist and is validated in validateFeatureExistence method
+			continue
+		}
+		policy := GetPolicyForStabilityLevel(stabilityLevel)
+		// immutability validation
+		if policy.Immutable && !featureRef.SkipStabilityPolicyValidation && policy.DefaultActivation != featureRef.Activate {
+			invalidFeatures.Insert(featureRef.Name)
+		} else if policy.VoidsWarranty && !featureRef.SkipStabilityPolicyValidation { // warranty validation
+			invalidFeatures.Insert(featureRef.Name)
+		}
+	}
+
+	if len(invalidFeatures) != 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"), r.Spec.Features,
+			fmt.Sprintf("cannot toggle features due to policy violation: %v."+
+				"Set feature.skipStabilityPolicyValidation to true to override policy."+
+				"Once set to true, cannot be set back to false.", invalidFeatures.List())))
+	}
+	return allErrors
+}
+
+// validateStabilityPolicyOverride validates stability policy override
+func (r *FeatureGate) validateStabilityPolicyOverride(oldObject *FeatureGate) field.ErrorList {
```

Comment:
Why are only 'oldObjects' validated? I'm assuming that an 'oldObject' is a config v1alpha1 CRD version FeatureGate. If my assumption is correct, why aren't 'new' objects using core v1alpha2 also checked for stability policy override validation? Or maybe this function is only for checking old objects? If so, maybe mention in the comment where the new objects are validated? I'm probably misunderstanding something.

Reply:
this webhook is for validating only core v1alpha2 resources.. here oldObject means the object before the update of the featuregate resource

Reply:
Ok, got it. Thanks.